proposals/LP0004-project-governance.md

```diff
@@ -0,0 +1,393 @@
+
+
+### Area Teams
+
+The second structure this proposal establishes is a collection of _area teams_. Members of _area teams_ are not required or expected to be experts in the area the team is responsible for. There are no skill or experience requirements to be on an _area team_.
+
+_Area teams_ have three main responsibilities.
+
+First, they are responsible for electing from among themselves a team secretary who will take notes of any team meetings and a team chair who represents the team on the _project council_.
+
+Second, _area teams_ are responsible for maintaining an up-to-date and comprehensive list of code owners for their area of the project. They can nominate any individual they deem appropriate as code owner of any area they are responsible for. The role of _code owner_ remains a volunteer role, and any individual can accept, decline, or resign the role for themselves as they feel appropriate.
+
+Finally, _area teams_ are responsible for facilitating decision making for their area of the project. Facilitating decision making can take any number of forms ranging from contributing to RFC discussions, helping mediate disagreements, or fulfilling roles originally delegated to Chris Lattner in the [LLVM Decision Making](https://github.com/llvm/llvm-www/blob/main/proposals/LP0001-LLVMDecisionMaking.md) process.
+
+When acting to facilitate decision making the _area team _should act as a mediator between different perspectives helping find common ground and recognizing that decisions need not be binary. The _area team_ should seek to find the best solution to the framed problem, which may not be any of the proposed alternatives. If agreement cannot be reached, the _area team_ may act as the final decision maker. In that capacity decisions of an _area team_ are considered final, but can be overruled by a 2/3 majority vote of the _project council_ or the _area team_ itself revisiting the issue.
+
+_Area teams_ are not intended to be direction setters or primary maintainers of their areas, although individuals on an _area team_ may fulfill that role separately. The _area team’s_ role is as a steward and moderator ensuring the health and smooth operation of the area.
+
+Each _area team_ will have 5 members elected by the _voting contributors_. Candidates for _area teams_ must be a _voting contributor_ and self-nominated. An individual cannot serve on two _area teams_.
+
+Members of an _area team_ are elected for 1 year terms. Individuals are encouraged to not run for more than two consecutive terms. To encourage turnover, incumbent members of an _area team_ must place in the top 4 candidates during an election. The fifth spot will always go to a non-incumbent candidate.
```

Comment:
This implies a requirement that there be at least one non-incumbent in every election slate. Who makes sure that happens?

(And who runs the elections generally? There's way too much passive voice in the whole election-process description.)

Reply:
I also find the restrictions of `5` and `always` to not be realistic over time.

We find it hard to get code owners to existing projects, and yet now we have N area teams with no overlap between them and mandate that we can always find a 6th person to replace one of the 5. This means we need `(5 + 1) * N` people, unless we `barrel shift` one person from each team onto the next, then we only really need `5 * N` people, which is still _a lot_!

Reply:
The current structure is 12 _area teams_ (maybe that is too many). With that number of teams and the requirement that one candidate be non-incumbent we basically need 72 people (60 elected roles + 12 new people to cycle in).

Of the ~1700 people who contributed to LLVM in the last 12 months, I desperately hope we can find 72 who are interested in project governance.

I constantly question whether it is hard to find code owners. We have lots of active contributors who aren't code owners. Given that code owners are really only responsible for ensuring that code gets reviewed (rather than actually reviewing it), it shouldn't be that hard to fill out the owners files more robustly.

I tend to think the problem is more that people don't self-nominate and we don't have anyone responsible for nominating. People not stepping down when they move on is also a big problem because the barrier to self-nominating or nominating someone else is _much_ higher if you're also trying to remove someone who isn't participating.

Reply:
Being a code owner is a thankless task. Many excellent developers with full qualifications to lead a particular area are horrible at being a code owner because they are not people persons or don't have the time to be the _last line of defense_.

Being a representative of an area team is less ungrateful (you actually have some tasks and recognition), but it's equally a people-person kind of task. We'll struggle to find 72 people that intersects those domains, and most likely will end up having more non-technical people in those teams than technical ones (random guessing, don't quote me on that).

On the nomination side... I'm not sure I'd like to be nominated (and then voted) in anything that I'm not comfortable. I'm not sure how that would work in practice.

Reply:
I'll clarify the wording, but nobody nominated against their will should be placed on the ballot.

Reply:
I think that if we make the terms longer than 1 year we're likely to end up with people disappearing mid-term or not running at all because a multi-year commitment is harder to sign up for.

@tstellar had suggested offsetting the terms for half the teams so that half of each team cycled every 6 months. I kinda like that idea myself, but I'm also unsure about the effort required to run elections every 6 months.

Reply:
The term limit requirement seems difficult to satisfy to me. The most active contributors have multi-year length periods of engagement.

Reply:
So someone can only be on an area team for 2 years?  That seems...aggressive. :-1:

Reply:
> This implies a requirement that there be at least one non-incumbent in every election slate.

Right. What happens when 3 people run for area team, and all 3 are incumbents? This document should clarify.

Reply:
> Of the ~1700 people who contributed to LLVM in the last 12 months, I desperately hope we can find 72 who are interested in project governance.

Even with that number of contributors, I still suspect for certain smaller areas we will fail to fill 5 volunteered seats. Then after 2 years we will end up cutting that down due to term limits.

---

@rnk and I were discussing in person that it seems helpful that the number is odd, so that the area team can't come to a stalemate in agreement.  Perhaps it's better to codify "an even number no greater than 5" than just "5" and mention this more explicitly in the doc.

Reply:
> So someone can only be on an area team for 2 years? That seems...aggressive. 👎

This is something I find a bit icky as well -- I think that in many cases, the set of people best placed to make a decision about an area of the code doesn't change all that much over the years, and this may force us to artificially select new members of the area team that are less suited to the task. At the same time, area teams are not dictators -- they are mediators in case the normal process fails to work, so I think this makes this a bit less relevant.

Personally, I'd be happy to drop the requirement/recommendation of turnover on area teams so that in case an area team is working well there is no need to replace members just for the sake of it.

Reply:
I agree with Nick and Louis here, I think we should drop the turnover, I think it's going to be hard to fill the spots anyway.